### PR TITLE
get Composite Modification endpoint

### PR DESCRIPTION
**Pull request by Mathieu-Deharbe**

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 1]
```diff
@@ -289,7 +295,7 @@ public Optional<NetworkModificationResult> duplicateModifications(UUID targetGro
     public Optional<NetworkModificationResult> insertCompositeModifications(UUID targetGroupUuid,
                                                                             UUID networkUuid, String variantId,
                                                                             ReportInfos reportInfos, List<UUID> modificationsUuids) {
-        List<ModificationInfos> modificationInfos = networkModificationRepository.getCompositeModificationsInfos(modificationsUuids);
+        List<ModificationInfos> modificationInfos = networkModificationRepository.getCompositeModificationsContentInfos(modificationsUuids, false);
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:298]: ContentInfos ? (content or infos have the same meaning) you could keep only Infos

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 2]
```diff
@@ -394,15 +394,28 @@ public List<ModificationInfos> getModificationsInfos(@NonNull List<UUID> uuids)
         return uuids.stream().map(entities::get).filter(Objects::nonNull).map(this::getModificationInfos).toList();
     }
 
+    /**
+     * @param onlyCommonData if true, only returns the basic data common to all the modifications. If false, returns complete modifications
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:398]: Could you explain the basic of the data and what do you mean by onlyCommonData? 
It doesn't make sense to me

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 3]
```diff
@@ -394,15 +394,28 @@ public List<ModificationInfos> getModificationsInfos(@NonNull List<UUID> uuids)
...
         uuids.forEach(uuid -> {
-            List<UUID> foundEntities = modificationRepository.findModificationIdsByCompositeModificationId(uuid);
-            List<ModificationInfos> orderedModifications = foundEntities
-                    .stream()
-                    .map(this::getModificationInfo)
-                    .toList();
+            List<UUID> networkModificationsUuids = modificationRepository.findModificationIdsByCompositeModificationId(uuid);
+            List<ModificationInfos> orderedModifications;
+            if (onlyCommonData) {
+                List<ModificationEntity> networkModifications = modificationRepository.findBaseDataByIdIn(networkModificationsUuids);
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:408]: networkModificationsEntities

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 4]
```diff
@@ -394,15 +394,28 @@ public List<ModificationInfos> getModificationsInfos(@NonNull List<UUID> uuids)
...
+                orderedModifications = networkModifications
+                        .stream()
+                        .map(this::getModificationInfos)
+                        .toList();
+            } else {
+                orderedModifications = networkModificationsUuids
+                        .stream()
+                        .map(this::getModificationInfo)
+                        .toList();
+            }
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:418]:
```
 if (uuids.isEmpty()) {
            return Collections.emptyList(); 
        }
            
```

and we can do better by fetching all modification uuids in one go like 
        
```Map<UUID, List<UUID>> networkModificationsMap = uuids.stream()
 .collect(Collectors.toMap(
                        uuid -> uuid,
                        modificationRepository::findModificationIdsByCompositeModificationId));
            }
```

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 5]
```diff
@@ -394,15 +394,28 @@ public List<ModificationInfos> getModificationsInfos(@NonNull List<UUID> uuids)
...
             entities.addAll(orderedModifications);
         }
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:420]:
```        
for (UUID uuid : uuids) {
            List<UUID> networkModificationsUuids = networkModificationsMap.get(uuid);
            if (onlyMetadata) {
                List<ModificationEntity> networkModifications = modificationRepository.findBaseDataByIdIn(networkModificationsUuids);
                entities.addAll(networkModifications.stream()
                        .map(this::getModificationInfos)
                        .toList());
            } else {
                entities.addAll(networkModificationsUuids.stream()
                        .map(this::getModificationInfo)
                        .toList());
            }
        }

        return Collections.unmodifiableList(entities); 
        }
```

**Review by ghazwarhili — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/modification/server/NetworkModificationController.java [hunk 6]
```diff
@@ -220,6 +220,18 @@ public ResponseEntity<UUID> createNetworkCompositeModification(@RequestBody List
         return ResponseEntity.ok().body(networkModificationService.createNetworkCompositeModification(modificationUuids));
     }
 
+    @GetMapping(value = "/network-composite-modification/{id}", produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/modification/server/NetworkModificationController.java:223]:
```suggestion
    @GetMapping(value = "/network-composite-modification/{uuid}", produces = MediaType.APPLICATION_JSON_VALUE)
```

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 7]
```diff
@@ -394,15 +394,28 @@ public List<ModificationInfos> getModificationsInfos(@NonNull List<UUID> uuids)
         return uuids.stream().map(entities::get).filter(Objects::nonNull).map(this::getModificationInfos).toList();
     }
 
+    /**
+     * @param onlyCommonData if true, only returns the basic data common to all the modifications. If false, returns complete modifications
+     * @return the data from all the network modification contained in the composite modification sent as parameters
+     */
     @Transactional(readOnly = true)
-    public List<ModificationInfos> getCompositeModificationsInfos(@NonNull List<UUID> uuids) {
+    public List<ModificationInfos> getCompositeModificationsContentInfos(@NonNull List<UUID> uuids, boolean onlyCommonData) {
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:402]: I'm not sure this method should return wether a list of modification base infos or a modification full infos. You could rather implement 2 two methods

File: src/main/java/org/gridsuite/modification/server/NetworkModificationController.java [hunk 8]
```diff
@@ -220,6 +220,18 @@ public ResponseEntity<UUID> createNetworkCompositeModification(@RequestBody List
...
+    @Operation(summary = "Get the content of a composite modification via its id")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "Content of the composite modification"),
+        @ApiResponse(responseCode = "404", description = "This composite modification does not exist")}
+    )
+    public ResponseEntity<List<ModificationInfos>> getCompositeModificationContent(@PathVariable("id") UUID compositeModificationId) {
```
[the remark this answers is not in the document]
Comment (Mathieu-Deharbe) [reply src/main/java/org/gridsuite/modification/server/NetworkModificationController.java:228]: For PathVariable `boolean required() default true`

If I add a true, intellij calls it a redundant parameter.

File: src/main/java/org/gridsuite/modification/server/NetworkModificationController.java [hunk 9]
```diff
@@ -220,6 +220,18 @@ public ResponseEntity<UUID> createNetworkCompositeModification(@RequestBody List
...
+        return ResponseEntity.ok()
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(networkModificationService.getCompositeModificationContentMetadata(compositeModificationId)
```
[the remark this answers is not in the document]
Comment (Mathieu-Deharbe) [reply src/main/java/org/gridsuite/modification/server/NetworkModificationController.java]: Renamed `getNetworkModificationsFromComposite`

I am still not very satisfied of it, but still better than metadata indeed.

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Mathieu-Deharbe) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java]: Ok but I don't think this is a good name : it sounds like we are getting the infos of the composite modifications but we are not. We get the infos of the network modifications that are inside the composite modifications and don't get those of the composite modifications themselves.

It sent me in the wrong direction when I was looking at functions.

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java (hunk 2 above)
Comment (Mathieu-Deharbe) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java]: All the network modifications use the `ModificationInfos`. The data inside this class is the common data to all the network modifications.

But they also have specific data extending this data like `BatteryModificationInfos` or `LineCreationInfos`.

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 10]
```diff
@@ -394,15 +394,28 @@ public List<ModificationInfos> getModificationsInfos(@NonNull List<UUID> uuids)
...
         List<ModificationInfos> entities = new ArrayList<>();
         uuids.forEach(uuid -> {
-            List<UUID> foundEntities = modificationRepository.findModificationIdsByCompositeModificationId(uuid);
-            List<ModificationInfos> orderedModifications = foundEntities
-                    .stream()
-                    .map(this::getModificationInfo)
-                    .toList();
+            List<UUID> networkModificationsUuids = modificationRepository.findModificationIdsByCompositeModificationId(uuid);
```
[the remark this answers is not in the document]
Comment (Mathieu-Deharbe) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java]: Do you mean renaming `networkModificationsUuids` ? But those are not entities.

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java (hunk 3 above)
Comment (Mathieu-Deharbe) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java]: OK.

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 11]
```diff
@@ -76,6 +76,12 @@ public List<ModificationInfos> getNetworkModifications(UUID groupUuid, boolean o
         return getNetworkModifications(groupUuid, onlyMetadata, errorOnGroupNotFound, false);
     }
 
+    @Transactional(readOnly = true)
+    // Need a transaction for collections lazy loading
+    public List<ModificationInfos> getCompositeModificationContentMetadata(UUID compositeModificationUuid) {
```
[the remark this answers is not in the document]
Comment (Mathieu-Deharbe) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: Renamed, removed all references to Metadata.

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 12]
```diff
@@ -76,6 +76,12 @@ public List<ModificationInfos> getNetworkModifications(UUID groupUuid, boolean o
         return getNetworkModifications(groupUuid, onlyMetadata, errorOnGroupNotFound, false);
     }
 
+    @Transactional(readOnly = true)
+    // Need a transaction for collections lazy loading
```
[the remark this answers is not in the document]
Comment (Mathieu-Deharbe) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: Only the comment or the @Transactional as well ?

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 1 above)
Comment (Mathieu-Deharbe) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: Renamed.

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 13]
```diff
@@ -394,6 +394,23 @@ public List<ModificationInfos> getModificationsInfos(@NonNull List<UUID> uuids)
         return uuids.stream().map(entities::get).filter(Objects::nonNull).map(this::getModificationInfos).toList();
     }
 
+    /**
+     * returns the data from all the network modifications contained in the composite modification sent as parameter
+     * but only returns the basic data common to all the modifications form the ModificationInfos, not from the extended classes
+     */
+    @Transactional(readOnly = true)
+    public List<ModificationInfos> getBasicNetworkModificationsFromComposite(@NonNull UUID uuid) {
+
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:403]: remove empty row

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 14]
```diff
@@ -394,6 +394,23 @@ public List<ModificationInfos> getModificationsInfos(@NonNull List<UUID> uuids)
...
+        List<UUID> networkModificationsUuids = modificationRepository.findModificationIdsByCompositeModificationId(uuid);
+
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:405]: remove empty row

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 15]
```diff
@@ -394,6 +394,23 @@ public List<ModificationInfos> getModificationsInfos(@NonNull List<UUID> uuids)
...
+        List<ModificationEntity> networkModificationsEntities = modificationRepository.findBaseDataByIdIn(networkModificationsUuids);
+
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:407]: remove empty row

**Review by etiennehomer — APPROVED** [review]

**Pull request merged**